transformer_engine/pytorch/fp8.py

```diff
@@ -1084,3 +1105,61 @@ def make_quantizers(self) -> list:
                 ]
             )
         )
+
+
+class CustomRecipeState(RecipeState):
+    """State for CustomRecipe: produce quantizers per tensor."""
+
+    recipe: CustomRecipe
+    mode: str
+    num_quantizers: int
+    device: Optional[torch.device]
+
+    def __init__(
+        self,
+        recipe: CustomRecipe,
+        *,
+        mode: str,
+        num_quantizers: int = 1,
+        device: Optional[torch.device] = None,
+    ) -> None:
+        self.recipe = recipe
+        self.mode = mode
+        self.num_quantizers = num_quantizers
+        if device is None:
+            device = torch.device("cuda")
+        self.device = device
+
+        if getattr(recipe, "qfactory", None) is None:
+            raise ValueError("CustomRecipe requires `qfactory`.")
+
+    def make_quantizers(self) -> list:
+        qfactory = self.recipe.qfactory
+        out = []
+        for i in range(self.num_quantizers):
+            # TODO(negvet): obtain forward_roles/backward_roles from module/op
+            # configuration (e.g., te.Linear uses (3/2), te.ops.Linear uses (2/1)).
+            # For now, hardcode the te.Linear layout
+            forward_roles = ("input", "weight", "output")
+            backward_roles = ("grad_output", "grad_input")
```

Comment:
Why pass the forward and backward roles separately into `_resolve_te_role` and `qfactory`? We can fully figure out the roles beforehand:
```suggestion
            roles = ["unknown"]
            if mode == "forward":
                roles = [("input", "weight", "output")[i % 3] for i in range(self.num_quantizers)]
            elif mode == "backward":
                roles = [("grad_output", "grad_input")[i % 2] for i in range(self.num_quantizers)]
```
No reason for the forward quantizers to know about backward roles, and vice versa. This is one reason I don't like the kwargs in `qfactory`. In the future, we can generalize the quantizer infrastructure so that each operation can pass in its own list of roles.